src/pro/javacard/gp/AccessControl.java

```diff
@@ -0,0 +1,769 @@
+			this.currentIndex = index;
+		}
+
+		public byte[] getData(){
+			return data;
+		}
+
+		public int getLength(){
+			return length;
+		}
+
+		public int getCurrentIndex(){
+			return currentIndex;
+		}
+	}
+
+	/**
+	 * Parse access rule list response.
+	 */
+	public static class AcrListResponse {
+
+		public List<RefArDo> acrList;
+
+		public AcrListResponse(List<RefArDo> acrList) {
+			this.acrList = acrList;
+		}
+
+		public static BerTlvData getAcrListData(BerTlvData previousData, byte[] data) throws GPDataException {
+
+			if (previousData == null &&
+					data.length > 2 &&
+					(data[0] == ACR_GET_DATA_RESP[0]) &&
+					(data[1] == ACR_GET_DATA_RESP[1])) {
+
+				int first = data[2] & 0xFF; // fist byte determining length
+				int length = 0; // actual length integer
+				int offset = 3; //offset
+
+				if (first < 0x80){
+					length = first & 0xFF;
+				}
+				else {
+					switch(first) {
+						case 0x81:
+							length = data[3] & 0xFF;
+							offset++;
+							break;
+						case 0x82:
+							length = ((data[3] & 0xFF) << 8) | (data[4] & 0xFF);
+							offset+=2;
+							break;
+						case 0x83:
+							length = ((data[3] & 0xFF) << 16) | ((data[4] & 0xFF) << 8) | (data[5] & 0xFF);
+							offset+=3;
+							break;
+						default:
+							throw new GPDataException("acr get data : bad BER TLV response format (GET_DATA)");
+					}
+				}
+				byte[] berData = new byte[length];
+				System.arraycopy(data, offset, berData, 0, data.length-offset);
+				return new BerTlvData(berData, length, data.length - offset);
+			}
+			else if (previousData != null) {
+				System.arraycopy(data, 0, previousData.getData(), previousData.currentIndex, data.length);
+				previousData.setCurrentIndex(data.length + previousData.currentIndex);
+				return previousData;
+			}
+			else {
+				throw new GPDataException("acr get data : bad response format (GET_DATA)");
+			}
+		}
+
+		public static AcrListResponse fromBytes(int length, byte[] data) throws GPDataException {
+			List<RefArDo> acrList = new ArrayList<>();
+
+			int offset = 0;
+
+			while (length > offset){
+				acrList.add(parseRefArDo(Arrays.copyOfRange(data, offset, data.length)));
+				offset += ((data[1 + offset] & 0xFF) + 2);
+			}
+			return new AcrListResponse(acrList);
+		}
+	}
+
+	/**
+	 * Parse REF_AR_DO object (p46 Secure Element Access Control v1.0).
+	 *
+	 * <p>
+	 * 0xE2 | length | REF-DO | AR-DO
+	 * </p>
+	 *
+	 * @param data REF_AR_DO data
+	 * @return
+	 * @throws GPDataException
+	 */
+	public static RefArDo parseRefArDo(byte[] data) throws GPDataException {
+		if (data.length > 2 &&
+				(data[0] == REF_AR_DO) &&
+				((data[1] & 0xFF) <= (data.length-2))) {
+
+			RefDo refDo = parseRefDo(Arrays.copyOfRange(data, 2, data.length));
+			ArDo arDo = parseArDo(Arrays.copyOfRange(data, 2 + (2 + (data[3] & 0xFF)), data.length));
+
+			return new RefArDo(refDo, arDo);
+		}
+		else {
+			throw new GPDataException("acr get data : bad response format (REF_AR_DO)");
```

Comment:
Regarding lowercase "acr" - what would be the overall name for the feature? ACR? SEAR? SEACR?

Reply:
I will use ACR